add additional services

phantomnat opened the pull request:
- getRemaining -> return up to 10 available seats
- reserve -> reserve the seat for up to X seconds
- confirm -> confirm booking
- cancel -> cancel the rsvp
add unit test
update apidoc

phantomnat committed:
add additional services
- getRemaining -> return up to 10 available seats
- reserve -> reserve the seat for up to X seconds
- confirm -> confirm booking
- cancel -> cancel the rsvp
add unit test
update apidoc

varokas commented:
thanks! this looks great. I'll play with it tonight my time and merge.

varokas commented on lib/book.js:
```diff
@@ -0,0 +1,182 @@
+        // reserve expire
+    } else {
+        // seatNo not valid or not available
+        // let get an available seat for user
```
For the new rule, we no longer need this logic to grab a seat. Let's make "seat" a required parameter for the "/book" API.

We will let the client refresh /bookings repeatedly to contend for resource. Just like real world =)

varokas commented on lib/book.js:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+    }
+    if (!(seatNo in seats)) {
```
again, no need for this. If the seatNo happens to reserved. we will simply returns false

varokas commented on index.js:
```diff
@@ -1,23 +1,44 @@
+app.get('/remaining/:noOfSeats*?', function(req, res) {
+    const result = book.getRemaining(req.params.noOfSeats)
+    res.json({ seats: result})
+})
```
to make this even simpler, we could just hard-coded number 10 as in rule. no need to have a parameter here.

varokas commented on index.js:
```diff
@@ -26,35 +47,85 @@ bookings.push({"round": currentRound, "seats": []})
-        currentCol = 1
-        currentRow = String.fromCharCode(currentRow.charCodeAt() + 1);
+app.post('/book', function(req, res) {
+    const result = book.reserve(req.body.seat)
```
See comment below about `reserve(seat)`. We should just reject request with `seat == null`

phantomnat replied:
`reserve()` can handle `null` and will return `false`

varokas commented:
@phantomnat - i tried it and this works really good. Thanks for your contribution. Let me know what you think about my comments. 

@mahasak please review (or add tossapon to help review)

tnuanchuay commented on lib/book.js:
```diff
@@ -0,0 +1,182 @@
+            }
+        }
+    }
+    if (!(seatNo in seats)) {
```
Should we go find the seat next round if there is no seat available in this round?
It seems like he/she need to confirm the last seat and then we can go further to next round.

We should have rules for this, if last seat isn't confirmed yet and customers are coming. they can/can't get next round seat. @varokas

varokas replied:
which is fine. we want to have some contention going on. but in light of this (https://github.com/bigbearsio/strength/issues/6#issuecomment-391597105) we probably will just remove rounds altogether. 

Let's skip this concern and merge this after other comments are addressed. Then we can do another PR to remove rounds?

tnuanchuay reviewed (approved):
Looks good to me, but have concern

phantomnat commented:
@tspn I think it would make sense if we're selling thing like 10K+ tickets for UEFA Champions League Final which is no advancing round.

I'll update the pr after we finish the discussion about this.

varokas commented:
@phantomnat  - sorry i forgot to submit review

phantomnat committed:
book requied seat no
get remaining always return up to 10 seats
update apidoc
update unit test
move unit test from ./spec to ./test

phantomnat commented:
lets check this again @varokas

varokas commented:
looks good. tested and works perfectly. Thanks a lot for contributing

varokas commented:
@mahasak @tspn  - im merging this
